examples/common/convert_c4.py

```diff
@@ -4,105 +4,392 @@
+                                         truncation=False,
+                                         padding=False)
+                for iids in encoded['input_ids']:
+                    buffer = buffer + bos_tokens + iids + eos_tokens
+                    while len(buffer) >= self.max_length:
+                        concat_sample = buffer[:self.max_length]
+                        buffer = buffer[self.max_length:]
+
+                        yield {
+                            # convert to bytes to store in MDS binary format
+                            'tokens': np.asarray(concat_sample).tobytes()
+                        }
+                        n_samples += 1
+                        if n_samples == self.expected_num_samples:
+                            return
+
+
+def build_hf_c4_dataset(split: str, mode: ConcatMode, max_length: Optional[int],
+                        bos_text: Optional[str], eos_text: Optional[str],
+                        tokenizer: Optional[PreTrainedTokenizerBase],
+                        expected_num_samples: int) -> IterableDataset:
     """Collect the samples for this dataset split.
 
     Args:
         split (str): Split name.
+        mode (ConcatMode): NO_CONCAT, CONCAT_TEXT, or CONCAT_TOKENS
+        bos_text (str): text to insert at the beginning of each sequence
+        eos_text (str): text to insert at the end of each sequence
+        tokenizer (PreTrainedTokenizerBase): if mode is CONCAT_TOKENS, the tokenizer to use
+        expected_num_samples (int): once we have generated this many samples, exit
 
     Returns:
         An IterableDataset.
     """
+    if mode == ConcatMode.NO_CONCAT:
+        dataset = ShardedC4(split, expected_num_samples=expected_num_samples)
+    elif mode == ConcatMode.CONCAT_TEXT:
+        dataset = ConcatC4(split=split,
+                           max_length=max_length,
+                           bos_text=bos_text,
+                           eos_text=eos_text,
+                           expected_num_samples=expected_num_samples)
+    else:
+        dataset = TokenizedC4(split=split,
+                              tokenizer=tokenizer,
+                              max_length=max_length,
+                              bos_text=bos_text,
+                              eos_text=eos_text,
+                              expected_num_samples=expected_num_samples)
+    return dataset
 
-    class ShardedC4(IterableDataset):
 
-        def __init__(self):
-            self.dataset = hf_datasets.load_dataset(path='c4',
-                                                    name='en',
-                                                    split=split,
-                                                    streaming=True)
+def _concat_sample_count(sample_count: int,
+                         mode: ConcatMode,
+                         max_length: Optional[int] = None) -> int:
+    emp_chars_per_sample = 2163  # measured for val split
```

Comment:
what is the implication of these estimates? does it mean that we might end up returning from `generate_samples` early if these are an underestimate?

Reply:
yes

Reply:
Is that a problem/why do we need to stop iteration at some sample number? can't we just exhaust the iterator?

Reply:
ohhh I don't need to do this, since I can keep track when building the buffers and concatenating...

Reply:
ok yeah, lets do that. I still don't actually understand why we need these caps at all though.

Reply:
I think we only actually want it for `train_small` ?

Reply:
`train` and `val` are the actual lengths. `train_small` is just an arbitrary amount. I made a change that uses the caps in a more reasonable way, only estimating to get tqdm a progress bar for the denominator